File: frontend/src/engine/tax/de/anlage-so.ts
/**
 * Anlage SO (Sonstige Einkuenfte) CSV generator.
 *
 * Ported from src/tax/de/anlage_so.py.
 *
 * Generates a CSV file formatted for use with German tax filing software
 * and Steuerberater (tax advisors).
 *
 * Structure:
 *   Section 1: Private Veraeusserungsgeschaefte (§23 EStG) - crypto disposals
 *   Section 2: Sonstige Einkuenfte (§22 Nr. 3 EStG) - staking rewards, etc.
 *
 * CRITICAL: All monetary values use decimal.js Decimal -- NEVER native number for money.
 */

import Decimal from 'decimal.js';
import type { Disposal, ReportFile } from '../../types';
import { HoldingPeriod } from '../../types';
import type { IncomeEvent } from '../us/income-report';

/** Format date as DD.MM.YYYY (German convention). */
function fmtDateDE(dt: string): string {
  try {
    const d = new Date(dt.slice(0, 10));
    if (isNaN(d.getTime())) return dt;
    const dd = String(d.getUTCDate()).padStart(2, '0');
    const mm = String(d.getUTCMonth() + 1).padStart(2, '0');
    const yyyy = d.getUTCFullYear();
    return `${dd}.${mm}.${yyyy}`;
  } catch {
    return dt;
  }
}

/** Format Decimal as German currency string: 1234,56 */
function fmtEur(amount: Decimal): string {
  return amount.toFixed(2).replace('.', ',');
}

function isShortTerm(d: Disposal): boolean {
  const hp = d.holdingPeriod;
  return ['short-term', 'short', 'SHORT_TERM'].includes(hp);
}

function isExempt(d: Disposal): boolean {
  return d.holdingPeriod === 'exempt' ||
    d.holdingPeriodEnum === HoldingPeriod.EXEMPT;
}

const TYPE_LABELS_DE: Record<string, string> = {
  reward: 'Staking Reward',
  airdrop: 'Airdrop',
  mining: 'Mining',
  validator: 'Validator',
  interest: 'DeFi-Zinsen',
};

/** Escape a semicolon-delimited field. */
function escField(val: string): string {
  if (val.includes(';') || val.includes('"') || val.includes('\n')) {
    return '"' + val.replace(/"/g, '""') + '"';
  }
  return val;
}

function buildSemicolonRow(fields: string[]): string {
  return fields.map(escField).join(';');
}

/**
 * Generate Anlage SO CSV.
 *
 * @param disposals    - Array of Disposal objects
 * @param incomeEvents - Array of income event objects
 * @param year         - Tax year
 * @returns            ReportFile with semicolon-delimited CSV content
 */
export function generateAnlageSO(
  disposals: Disposal[],
  incomeEvents: IncomeEvent[],
  year: number,
): ReportFile {
  const lines: string[] = [];

  // --- Section 1: Private Veraeusserungsgeschaefte (§23 EStG) ---
  lines.push(escField('Anlage SO - Private Veraeusserungsgeschaefte (Kryptowaehrungen)'));
  lines.push(escField(`Steuerjahr: ${year}`));
  lines.push('');

  lines.push(buildSemicolonRow([
    'Zeile',
    'Art des Wirtschaftsguts',
    'Anschaffungsdatum',
    'Veraeusserungsdatum',
    'Veraeusserungspreis (EUR)',
    'Anschaffungskosten (EUR)',
    'Gewinn/Verlust (EUR)',
    'Haltefrist',
    'Steuerpflichtig',
  ]));

  let zeile = 1;
  let totalTaxableGain = new Decimal('0');
  let totalExemptGain = new Decimal('0');

  for (const d of disposals) {
    const dy = parseInt(d.date.slice(0, 4), 10);
    if (dy !== year) continue;

    const short = isShortTerm(d);
    const exempt = isExempt(d);
    const haltefrist = short ? '<=365 Tage' : '>365 Tage (steuerfrei)';
    const steuerpflichtig = short ? 'Ja' : 'Nein';

    if (short) {
      totalTaxableGain = totalTaxableGain.plus(d.gainLossUsd);
    } else {
      totalExemptGain = totalExemptGain.plus(d.gainLossUsd);
    }

    const description =
      `${d.amount.toFixed(8).replace(/0+$/, '').replace(/\.$/, '')} ${d.token}`;

    lines.push(buildSemicolonRow([
      String(zeile),
      description,
      'Verschiedene',
      fmtDateDE(d.date),
      fmtEur(d.proceedsUsd),
      fmtEur(d.costBasisUsd),
      fmtEur(d.gainLossUsd),
      haltefrist,
      steuerpflichtig,
    ]));
    zeile++;
  }

  lines.push('');
  lines.push(buildSemicolonRow([
    '', 'Summe steuerpflichtig', '', '', '', '', fmtEur(totalTaxableGain),
  ]));
  lines.push(buildSemicolonRow([
    '', 'Summe steuerfrei (Spekulationsfrist)', '', '', '', '', fmtEur(totalExemptGain),
  ]));

  // --- Section 2: Sonstige Einkuenfte (§22 Nr. 3 EStG) ---
  lines.push('');
  lines.push('');
  lines.push(escField('Sonstige Einkuenfte aus Kryptowaehrungen (\u00a722 Nr. 3 EStG)'));
  lines.push('');
  lines.push(buildSemicolonRow([
    'Zeile', 'Art der Einkuenfte', 'Datum', 'Token', 'Menge', 'Wert (EUR)',
  ]));

  let totalIncome = new Decimal('0');
  let incomeZeile = 1;

  for (const e of incomeEvents) {
    const dateYear = e.date.slice(0, 4);
    if (dateYear !== String(year)) continue;

    const usdVal = new Decimal(e.usdValue || '0');
    totalIncome = totalIncome.plus(usdVal);

    lines.push(buildSemicolonRow([
      String(incomeZeile),
      TYPE_LABELS_DE[e.txType] ?? e.txType,
      fmtDateDE(e.date),
      e.token,
      e.amount,
      fmtEur(usdVal),
    ]));
    incomeZeile++;
  }

  lines.push('');
  lines.push(buildSemicolonRow([
    '', 'Summe Sonstige Einkuenfte', '', '', '', fmtEur(totalIncome),
  ]));

  return {
    filename: `anlage_so_${year}.csv`,
    content: lines.join('\n') + '\n',
    mimeType: 'text/csv',
    reportType: 'anlage_so',
  };
}

File: src/tax/de/anlage_so.py
"""
Anlage SO (Sonstige Einkuenfte) CSV generator.

Generates a CSV file formatted for use with German tax filing software
and Steuerberater (tax advisors).

Structure:
  Section 1: Private Veraeusserungsgeschaefte (§23 EStG) - crypto disposals
  Section 2: Sonstige Einkuenfte (§22 Nr. 3 EStG) - staking rewards, etc.

Columns for disposals:
  Zeile, Art des Wirtschaftsguts, Anschaffungsdatum, Veraeusserungsdatum,
  Veraeusserungspreis, Anschaffungskosten, Gewinn/Verlust
"""
from __future__ import annotations

import csv
import io
from datetime import datetime
from decimal import Decimal

from src.tax.models import HoldingPeriod


def _fmt_date_de(dt) -> str:
    """Format date as DD.MM.YYYY (German convention)."""
    if isinstance(dt, datetime):
        return dt.strftime("%d.%m.%Y")
    if isinstance(dt, str):
        try:
            return datetime.fromisoformat(dt[:10]).strftime("%d.%m.%Y")
        except (ValueError, TypeError):
            return dt
    if hasattr(dt, "strftime"):
        return dt.strftime("%d.%m.%Y")
    return str(dt)


def _fmt_eur(amount: Decimal) -> str:
    """Format Decimal as German currency string: 1234,56"""
    return f"{amount:.2f}".replace(".", ",")


def _is_short_term(d) -> bool:
    hp = d.holding_period
    if isinstance(hp, HoldingPeriod):
        return hp == HoldingPeriod.SHORT_TERM
    return str(hp) in ("short-term", "short", "SHORT_TERM")


def generate_anlage_so_csv(
    disposals: list,
    income_events: list,
    year: int,
) -> str:
    """
    Generate Anlage SO CSV.

    Args:
        disposals: list of Disposal objects
        income_events: list of income event dicts/objects
        year: tax year

    Returns:
        CSV string in Anlage SO format
    """
    output = io.StringIO()
    writer = csv.writer(output, delimiter=";")

    # --- Section 1: Private Veraeusserungsgeschaefte (§23 EStG) ---
    writer.writerow(["Anlage SO - Private Veraeusserungsgeschaefte (Kryptowaehrungen)"])
    writer.writerow([f"Steuerjahr: {year}"])
    writer.writerow([])

    writer.writerow([
        "Zeile",
        "Art des Wirtschaftsguts",
        "Anschaffungsdatum",
        "Veraeusserungsdatum",
        "Veraeusserungspreis (EUR)",
        "Anschaffungskosten (EUR)",
        "Gewinn/Verlust (EUR)",
        "Haltefrist",
        "Steuerpflichtig",
    ])

    zeile = 1
    total_taxable_gain = Decimal("0")
    total_exempt_gain = Decimal("0")

    for d in disposals:
        if not hasattr(d.date, "year") or d.date.year != year:
            continue

        is_short = _is_short_term(d)
        hp = d.holding_period
        if isinstance(hp, HoldingPeriod):
            is_exempt = hp == HoldingPeriod.EXEMPT
        else:
            is_exempt = str(hp) == "exempt"

        haltefrist = "<=365 Tage" if is_short else ">365 Tage (steuerfrei)"
        steuerpflichtig = "Ja" if is_short else "Nein"

        if is_short:
            total_taxable_gain += d.gain_loss_usd
        else:
            total_exempt_gain += d.gain_loss_usd

        description = f"{d.amount:.8f} {d.token}".rstrip("0").rstrip(".")

        writer.writerow([
            zeile,
            description,
            "Verschiedene",  # VARIOUS - multiple lot dates
            _fmt_date_de(d.date),
            _fmt_eur(d.proceeds_usd),
            _fmt_eur(d.cost_basis_usd),
            _fmt_eur(d.gain_loss_usd),
            haltefrist,
            steuerpflichtig,
        ])
        zeile += 1

    writer.writerow([])
    writer.writerow(["", "Summe steuerpflichtig", "", "", "", "", _fmt_eur(total_taxable_gain)])
    writer.writerow(["", "Summe steuerfrei (Spekulationsfrist)", "", "", "", "", _fmt_eur(total_exempt_gain)])

    # --- Section 2: Sonstige Einkuenfte (§22 Nr. 3 EStG) ---
    writer.writerow([])
    writer.writerow([])
    writer.writerow(["Sonstige Einkuenfte aus Kryptowaehrungen (\u00a722 Nr. 3 EStG)"])
    writer.writerow([])
    writer.writerow([
        "Zeile",
        "Art der Einkuenfte",
        "Datum",
        "Token",
        "Menge",
        "Wert (EUR)",
    ])

    total_income = Decimal("0")
    zeile = 1

    for e in income_events:
        if isinstance(e, dict):
            e_date = e.get("date", "")
            e_type = e.get("tx_type", "")
            e_token = e.get("token", "")
            e_amount = e.get("amount", "0")
            e_usd = e.get("usd_value", Decimal("0"))
        else:
            e_date = getattr(e, "date", "")
            e_type = getattr(e, "tx_type", "")
            e_token = getattr(e, "token", "")
            e_amount = getattr(e, "amount", "0")
            e_usd = getattr(e, "usd_value", Decimal("0"))

        date_str = str(e_date)[:4]
        if date_str != str(year):
            continue

        usd_val = Decimal(str(e_usd)) if not isinstance(e_usd, Decimal) else e_usd
        total_income += usd_val

        _TYPE_LABELS = {
            "reward": "Staking Reward",
            "airdrop": "Airdrop",
            "mining": "Mining",
            "validator": "Validator",
            "interest": "DeFi-Zinsen",
        }

        writer.writerow([
            zeile,
            _TYPE_LABELS.get(e_type, str(e_type)),
            _fmt_date_de(e_date),
            e_token,
            str(e_amount),
            _fmt_eur(usd_val),
        ])
        zeile += 1

    writer.writerow([])
    writer.writerow(["", "Summe Sonstige Einkuenfte", "", "", "", _fmt_eur(total_income)])

    return output.getvalue()
